tremor-script/src/ast.rs

```diff
@@ -1363,10 +1751,40 @@ pub enum Pattern<'script> {
 impl<'script> Pattern<'script> {
     fn is_default(&self) -> bool {
         matches!(self, Pattern::Default)
+            || matches!(self, Pattern::DoNotCare)
+            || if let Pattern::Assign(AssignPattern { pattern, .. }) = self {
+                pattern.as_ref() == &Pattern::DoNotCare
+            } else {
+                false
+            }
     }
     fn is_assign(&self) -> bool {
         matches!(self, Pattern::Assign(_))
     }
+    fn is_exclusive_to(&self, other: &Self) -> bool {
+        match (self, other) {
+            // Two literals that are different are distinct
+            (Pattern::Expr(ImutExprInt::Literal(l1)), Pattern::Expr(ImutExprInt::Literal(l2))) => {
+                l1 != l2
+            }
+            // For record patterns we compare directly
+            (Pattern::Record(r1), Pattern::Record(r2)) => {
+                r1.is_exclusive_to(r2) || r2.is_exclusive_to(r1)
```

Comment:
So we can implement exclusivity in one direction instead of both, for example for `extractors` and `==` we only implement the `does an extractor is exclusive to a ==` not `is a == exclusive to an extractor`.

That said I'm not 100% sure where the 'right' place is to do this doubling :( it is in more places then I liked

Reply:
i think it is ok. We can revisit this when needed. I cannot come up with a case where the exclusivity relation is not commutative. So if anyhow, this might only be a problem of doing something twice, but thats ok as we are not in the hot path.